### PR TITLE
First set of changes to enable Solace JMS API support

### PR DESCRIPTION
**Pull request by mzhang444**
## Fixes bug
Add implements Serializable to JTBProperty.java in order to fix NotSeriablizableException when doing JTBMessageTemplate.deepClone(). Otherwise, if a message template defines custom Properties, when trying to execute send script from Script Center, exception would be thrown.
## New Features
Add support for Solace JMS API
## Description of proposed Change
  -
  -
  -

### CODE REVIEW

**Comment by titou10titou10** [comment]
First, thank for the contribution!

**TL; DR;**
I will not accept the code change in`"JTBConnection"`and`"QManager"`as the browsing problem has to be addressed in the Solace JMS client, not in JMSToolBox

**Why?**
After a quick review of the code, changing the way the messages are "browsed" in JTB is definitively not the way to go
With JMS (And so in JMSToolBox), browsing message is done with this standard code:

    QueueBrowser browser = jmsSession.createBrowser(...) {
    Enumeration<?> msgs = browser.getEnumeration();    
    while (msgs.hasMoreElements()) {    
       Message message = (Message) msgs.nextElement();    
      ...
    }

As already discussed with Tony, it appears that the implementation of the `"browser.getEnumeration()"` method within the JMS Solace client is broken (as stated in a comment in one JMS sample class).
In my preliminary tests, it appeared that `"msgs.hasMoreElements()"`always return`"false"`on first call even if there are messages on the queue. Also `"msgs.nextElement()"`can return`"nul"l` if no messages are present which is not allowed by the Java specs!
As stated in the Java `Enumeration`interface  here: 
https://docs.oracle.com/javase/8/docs/api/java/util/Enumeration.html
`"hasMoreElements()"`returns an element or throws a `NoSuchElementException`if none are present and `hasMoreElements`must be consistent with the above

This behavior is bad, not only for JMSToolBox, but with all Java/JMS clienst that will try to browse messages.,

So the solution is to fix the probem in the Solace JMS client, maybe by implementing the logic you coded here in the Solace JMS Client. So "browsing" messages in a JMS way will work for all JMS clients including JMSToolBox

**Comment by titou10titou10** [comment]
Monica,     
In the`SolaceQManager.`class, you implemented 2 ways to retrieve the destinations. One via JNDI based on some JNDI Destination names defined in the session. This is not in the philosophy of JTB to do that (ie list the destination names in the session def..) so I'll not keep this 

The second way uses some kind of management interface. Great! 
The classes for this interface come from the `"solace-samples-semp.jar"` jar. What is that jar? Is this a jar that is freely available, or available to Solace customers only? Is it "production ready" or something that is being currently developed for JTB?

Is that "management interface" something that is currently available in Solace? where is it documented? How is it configured? etc

In the code you added 2 parameters for`mgmtUsername`and`mgmtPassword`. is it necessary? should'nt it be the same user/password that the one to get the JMS Connection?

Thx
(BTW I rebased the solace branch on the current dev branch)

**Comment by mzhang444** [comment]
Hi, Denis. Thanks for looking at the pull request.

In our broker, messaging interface and management interface are separate.  Therefore, they have different URL, username and password. JMS clients and JNDI lookup are over messaging interface, while SEMP users are over management interface. Depending on the network configuration, users may have access to the messaging interface, but be blocked on management interface due to firewall. That's why we add two kind of lookup mechanisms just in case the user cannot access the management interface.

Unfortunately, our JNDI lookup implementation only supports lookup one object at a time, for first approach,  parameters with names "jndi*" are used to pass in a list of JNDI names so that SolaceQManager can do look up one at a time. If this approach is removed, without management interface access, user won't be able to see any queues or topics.

The second approach uses Solace management protocol SEMP. solace-samples-semp.jar and the other dependencies you saw are Java client built from the public swagger doc using swagger-codegen-cli. It was built by following instructions from https://github.com/SolaceSamples/solace-samples-semp.
The samples are widely used by people visiting Solace web site. Here is the link to Solace samples:
https://solace.com/samples/solace-samples-semp/

The second approach doesn't require specify destination names by default. The optional physicalQueueName parameter is there so that user can pick a specific queue to work with.

Also, our management interface requires paging to showing a collection of queues because a broker can support thousands of queues. Since I cannot find any paging support in the toolbox UI, I made discoverDestinations return the first 100. physicalQueueName parameter is there so that user can pick a queue that is not in the 100 entries. 

Thanks.
Monica

**Comment by titou10titou10** [comment]
Based on your work, I just commit a working version of the Solace plugin usingSEMP
There is till work to do (Q + Topic infos etc, SSL connection, help etc)
Q: Is there a REST verb to empty a Queue in one command, instead of iterating on all messages present in the Q to "consume" them?
Thx

**Comment by titou10titou10** [comment]
I just commited a cleaner/more complete version of the plugin
You'll have to delete/recreate your JTB session
Please review it, and check if some other parameters shoud lbe added on the connection factory and/or for semp
TODO: 
- SSL for JMS Connection and SSL for SEMP
- Test in various scenarios
- WiKi doc, including the requirements on the server to use the plugin (ie user/roles/setup)
- When it will be available, use the future SEMP feature to retrieve Q depth and empty a Q in an efficient way

**Comment by mzhang444** [comment]
> Based on your work, I just commit a working version of the Solace plugin usingSEMP
> There is till work to do (Q + Topic infos etc, SSL connection, help etc)
> Q: Is there a REST verb to empty a Queue in one command, instead of iterating on all messages present in the Q to "consume" them?
> Thx

**Comment by mzhang444** [comment]
> Q: Is there a REST verb to empty a Queue in one command, instead of iterating on all messages present in the Q to "consume" them?

We don't have REST support to empty whole Queue at this moment, but it will be added in future release. We do have REST support to delete a particular message by its Solace SMF messageId, not JMS Message ID.

**Comment by mzhang444** [comment]
HI, Denis. I have a few comments regarding your commit:
1. Discovering destinations using SEMP only assume jmstoolbox user has access to both messaging interface and management interface. If their network configuration doesn't allow management interface access, then they cannot use the tool.
2. The list of topics discovered are not JMS Topics, but Topic Endpoints in Solace message broker. JMS Topic is not Topic Endpoint. Topic Endpoint is an endpoint that attracts guaranteed messages matching the topic that a JMS client subscribes to.
3. In org.titou10.jtb.build/pom.xml, "<jtb.jre11.64>" points to hardcoded location, can you change it to ${env.JAVA_HOME}
4. Create a message Template with Properties, then create a script from Script Centre, then use that template to send messages, NullPointerException occurs due to JTBProperty.java not Serializable

Monica

**Comment by titou10titou10** [comment]
Monica,
Could you send me a message with your email adress to continue this discussion and receive my answers to titou10.titou10@gmail.com
Thx
